crates/bevy_render/src/batching/mod.rs

```diff
@@ -25,65 +29,107 @@ pub struct NoAutomaticBatching;
 ///   result of the `prepare_and_batch_meshes` system, e.g. due to having to split
 ///   data across separate uniform bindings within the same buffer due to the
 ///   maximum uniform buffer binding size.
-pub struct BatchMeta<T: PartialEq> {
+struct BatchMeta<T: PartialEq> {
     /// The pipeline id encompasses all pipeline configuration including vertex
     /// buffers and layouts, shaders and their specializations, bind group
     /// layouts, etc.
     pub pipeline_id: CachedRenderPipelineId,
     /// The draw function id defines the RenderCommands that are called to
     /// set the pipeline and bindings, and make the draw command
     pub draw_function_id: DrawFunctionId,
-    pub material_bind_group_id: Option<T>,
-    pub mesh_asset_id: AssetId<Mesh>,
     pub dynamic_offset: Option<NonMaxU32>,
+    pub user_data: T,
 }
 
 impl<T: PartialEq> PartialEq for BatchMeta<T> {
     #[inline]
     fn eq(&self, other: &BatchMeta<T>) -> bool {
         self.pipeline_id == other.pipeline_id
             && self.draw_function_id == other.draw_function_id
-            && self.mesh_asset_id == other.mesh_asset_id
             && self.dynamic_offset == other.dynamic_offset
-            && self.material_bind_group_id == other.material_bind_group_id
+            && self.user_data == other.user_data
     }
 }
 
+pub trait GetBatchData {
+    type Query: ReadOnlyWorldQuery;
+    type CompareData: PartialEq;
+    type BufferData: GpuArrayBufferable + Sync + Send + 'static;
+    fn get_batch_data(
+        batch_data: <Self::Query as WorldQuery>::Item<'_>,
```

Comment:
Hmm, what should this anonymous lifetime be? Seems a bit sus in an API?

Reply:
I guess it doesn't matter as it's only borrowed for the duration of get_batch_data, for now?